src/org/jetbrains/plugins/scala/caches/HydraArtifactsCache.scala

```diff
@@ -0,0 +1,43 @@
+package org.jetbrains.plugins.scala.caches
+
+import java.util.concurrent.ConcurrentHashMap
+
+import com.intellij.openapi.diagnostic.Logger
+import org.jetbrains.plugins.scala.project.Platform
+import org.jetbrains.plugins.scala.project.template.Downloader
+
+/**
+  * @author Maris Alexandru
+  */
+object HydraArtifactsCache {
+  private val LOG = Logger.getInstance(this.getClass)
+  private val cache = new ConcurrentHashMap[(String, String), Seq[String]]()
+  private val SplitRegex = "\\[info\\] \\* Attributed\\(|\\)"
+  private val ArtifactsRegex = "\\[info\\] \\* Attributed\\(.*\\)"
+  private val HydraRegex = ".*hydra_(\\d+\\.\\d+\\.\\d+)-(\\d+\\.\\d+\\.\\d+(-SNAPSHOT)?)\\.jar".r
+
+  private def cacheArtifacts(artifacts: String): Unit = {
+    val paths = artifacts.split("\n").filter(s => s.matches(ArtifactsRegex)).map(s => s.split(SplitRegex)(1))
+    val hydraPath = paths.filter(s => s.matches(HydraRegex.regex))(0)
+    val HydraRegex(scalaVersion, hydraVersion, _) = hydraPath
+    val hydraBridge = s"${System.getProperty("user.home")}/.ivy2/cache/com.triplequote/hydra-bridge_1_0/srcs/hydra-bridge_1_0-$hydraVersion-sources.jar"
+    cache.put((scalaVersion, hydraVersion), hydraBridge +: paths)
+  }
+
+  def getOrDownload(scalaVersion: String, hydraVersion: String,listener: (String) => Unit) = {
+    val artifacts = cache.getOrDefault((scalaVersion, hydraVersion), Seq.empty)
+
+    if(artifacts.isEmpty) {
+      val stringBuilder = new StringBuilder
+      Downloader.downloadScala(Platform.Hydra, s"${scalaVersion}_$hydraVersion", (text: String) => {
+        stringBuilder.append(text)
+        listener(text)
+      })
+      LOG.info(stringBuilder.toString())
+      cacheArtifacts(stringBuilder.toString)
+      cache.get((scalaVersion, hydraVersion))
```

Comment:
Here we could have a race condition, i.e., the value stored in the cache when calling `cacheArtifacts` might be different from the value read by `cache.get((scalaVersion, hydraVersion))`. To fix this, please change the return type of `cacheArtifacts` to return a tuple2 with the `scalaVersion` and the `hydraVersion`, so that you don't need to access the `cache` here.